### ISSUE TITLE
TypeScript Declarations

### ISSUE THREAD

**Issue by dillonredding** [digitalbazaar/vc-status-list#17]
I'm looking to use this library in a TypeScript app, but there are no type declarations. I don't see much (if any) TypeScript in digitalbazaar's repos, so I thought I'd ask here: what is the preference for contributing [type declarations](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html)?

For those unfamiliar with TypeScript, there are a few options when publishing types (in order of preference):
1. Write the code in TypeScript.
   * Pro: Types are easily [generated](https://www.typescriptlang.org/tsconfig#declaration)
   * Con: Adds a build step before publishing to NPM
1. Use [JSDoc to specify](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) and [generate types](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html).
   * Pro: Keeps the codebase in JavaScript
   * Con: Adds a build step before publishing to NPM
1. Publish types independently to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped).
   * Pro: Codebase remains unchanged
   * Cons: Types are manually maintained and can easily be incorrect or out of date

**Comment by cre8** [comment]
To open this question again, I think option 3 is the best. We don't need to change the existing code or break anything.

**Comment by davidlehn** [comment]
We write our code in standard ECMAScript/JavaScript and don't want to include TypeScript toolchains in our repos to generate type files.  I'd suggest you explore option 3.  If those external tools use the JSDocs here, maybe that is a low effort solution?

**Comment by cre8** [comment]
Just for the people needing the types right now:
statuslist.d.ts
```
declare module '@digitalbazaar/vc-status-list' {
  export class StatusList {
    constructor(options?: { length?: number; buffer?: ArrayBuffer });

    setStatus(index: number, status: boolean): void;
    getStatus(index: number): boolean;
    encode(): Promise<ArrayBuffer>;
    static decode({
      encodedList,
    }: {
      encodedList: ArrayBuffer;
    }): Promise<StatusList>;
  }
}
```
vc-status-list.d.ts
```
declare module '@digitalbazaar/vc-status-list' {
  import { StatusList } from './statuslist';

  export async function createList({
    length,
  }: {
    length: number;
  }): Promise<StatusList>;

  export async function decodeList({
    encodedList,
  }: {
    encodedList: any;
  }): Promise<StatusList>;

  export async function createCredential({
    id,
    list,
    statusPurpose,
  }: {
    id: string;
    list: StatusList;
    statusPurpose: string;
  }): Promise<any>;

  export async function checkStatus({
    credential,
    documentLoader,
    suite,
    verifyStatusListCredential,
    verifyMatchingIssuers,
  }?: {
    credential: any;
    documentLoader: (iri: string) => Promise<any>;
    suite?: any;
    verifyStatusListCredential?: boolean;
    verifyMatchingIssuers?: boolean;
  }): Promise<{ verified: boolean; error?: any }>;

  export function statusTypeMatches({
    credential,
  }: {
    credential: any;
  }): boolean;

  export function assertStatusList2021Context({
    credential,
  }: {
    credential: any;
  }): void;

  export function getCredentialStatus({
    credential,
    statusPurpose,
  }: {
    credential: any;
    statusPurpose: 'revocation' | 'suspension';
  }): any;

  // Add more type declarations for the functions and utilities here

  export function isArrayOfObjects(x: any): boolean;
}
```

Place them in a folder and add it in the tsconfig: `"typeRoots": ["node_modules/@types", "types"]`

**Comment by GHkrishna** [comment]
Hello @cre8 I tried implementing the suggested change for ts but this gives an error:

```
const list = (0, vc_status_list_1.createList)({ length: length });
                                                     ^

TypeError: (0 , vc_status_list_1.createList) is not a function
    at StatusController.createBitStringStatusList
```
Not sure if the issue is with config file or something else.

**Comment by cre8** [comment]
> Hello @cre8 I tried implementing the suggested change for ts but this gives an error:
> 
> ```
> const list = (0, vc_status_list_1.createList)({ length: length });
>                                                      ^
> 
> TypeError: (0 , vc_status_list_1.createList) is not a function
>     at StatusController.createBitStringStatusList
> ```
> 
> Not sure if the issue is with config file or something else.

It can be that the types are not compatible with the code anymore since they changed it after my comment

**Comment by GHkrishna** [comment]
> It can be that the types are not compatible with the code anymore since they changed it after my comment

I don't think this shall be an issue because the declaratives that I am using are generated according to the latest code on the main branch


Below is the initial issue I'm getting upon running my app(compilation succeds).
``` 
const vc_status_list_1 = require("@digitalbazaar/vc-status-list");
                         ^

Error [ERR_REQUIRE_ESM]: require() of ES Module /path/node_modules/@digitalbazaar/vc-status-list/lib/index.js from /some-path/build/again-some-path/myCompiledFile.js not supported.
Instead change the require of index.js in some-path/build/again-some-path/myCompiledFile.js to a dynamic import() which is available in all CommonJS modules.
```

To resolve this when I make the following change as suggested by the message(in the compiled js file), the code looks something like this
```
const vc_status_list_1 = import("@digitalbazaar/vc-status-list");
```

With this change I'm able to run the application but the response I am getting is as follows,
```
"(0 , vc_status_list_1.createList) is not a function"
```

Following are some changes that I've made,
- "typeRoots": ["node_modules/@types", "src/types"]
(this successfully picks the types)
- Created new declarative files with the updated code and placed them as suggested

**Comment by davidlehn** [comment]
- [`import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) returns a promise.  You probably want `await import(...)`.
- We added a status note to the README for this project.  You should look at https://github.com/digitalbazaar/vc-bitstring-status-list / https://github.com/w3c/vc-bitstring-status-list for future use.  It's still a work-in-progress.  That lib is also plain ESM with no special TypeScript declarations.
